src/synthcity/plugins/generic/plugin_ddpm.py

```diff
@@ -0,0 +1,193 @@
+            scheduler=scheduler,
+            device=device,
+            verbose=verbose,
+            log_interval=log_interval,
+            print_interval=print_interval,
+            model_type=model_type,
+            rtdl_params=rtdl_params,
+            dim_label_emb=dim_label_emb,
+            n_iter_min=n_iter_min,
+            n_iter_print=n_iter_print,
+            patience=patience,
+        )
+
+    @staticmethod
+    def name() -> str:
+        return "ddpm"
+
+    @staticmethod
+    def type() -> str:
+        return "generic"
+
+    @staticmethod
+    def hyperparameter_space(**kwargs: Any) -> List[Distribution]:
+        """
+        Hyperparameter Search space (from the paper)
+        ----------------------------------------------
+        Learning rate           LogUniform[0.00001, 0.003]
+        Batch size              Cat{256, 4096}
+        Diffusion timesteps     Cat{100, 1000}
+        Training iterations     Cat{5000, 10000, 20000}
+        Number of MLP layers    Int{2, 4, 6, 8}
+        MLP width of layers     Int{128, 256, 512, 1024}
+        Proportion of samples   Float{0.25, 0.5, 1, 2, 4, 8}
+        ----------------------------------------------
+        Dropout                 0.0
+        Scheduler               cosine (Nichol, 2021)
+        Gaussian diffusion loss MSE
+        """
+        return [
+            # TODO: change to loguniform distribution
+            CategoricalDistribution(name="lr", choices=[1e-5, 1e-4, 1e-3, 2e-3, 3e-3]),
+            CategoricalDistribution(name="batch_size", choices=[256, 4096]),
+            CategoricalDistribution(name="num_timesteps", choices=[100, 1000]),
+            CategoricalDistribution(name="n_iter", choices=[5000, 10000, 20000]),
+            CategoricalDistribution(name="num_layers", choices=[2, 4, 6, 8]),
+            CategoricalDistribution(name="dim_hidden", choices=[128, 256, 512, 1024]),
+        ]
+
+    def _fit(self, X: DataLoader, *args: Any, **kwargs: Any) -> "TabDDPMPlugin":
+        cond = None
+        if args:
+            if len(args) > 1:
+                raise ValueError("Only one positional argument is allowed")
+            if "cond" in kwargs:
+                raise ValueError("cond is already given by the positional argument")
+            cond = args[0]
+        elif "cond" in kwargs:
+            cond = kwargs.pop("cond")
+
+        if self.is_classification:
+            if cond is not None:
+                raise ValueError(
+                    "cond is already given by the labels for classification"
+                )
+            _, cond = X.unpack()
+            self._labels, self._cond_dist = np.unique(cond, return_counts=True)
+            self._cond_dist = self._cond_dist / self._cond_dist.sum()
+
+        # NOTE: should we include the target column in `df`?
+        df = X.dataframe()
+
+        if cond is not None:
+            cond = pd.Series(cond, index=df.index)
+
+        # self.encoder = TabularEncoder().fit(X)
```

Comment:
Why are you not using the tabular encoding?

Reply:
Because the Gaussian multinomial diffusion module needs to know the number of classes for each discrete feature before it applies `torch.nn.functional.one_hot` on these features, and it also preprocesses the continuous features differently.

Reply:
OK, then remove this comment, and leave a comment in the code explaining why the TabularEncoding is not needed.